### PR TITLE
`builtins.sum`: Items in the iterable must support addition with `int` if no `start` value is given

### LINKED ISSUE #7574: `sum` stub too permissive
Currently the stub for the `sum` built-in allows any iterable as an argument
https://github.com/python/typeshed/blob/ae6ff79c0e14cd054bf07374e7091a098a46cf1f/stdlib/builtins.pyi#L1480-L1489

This means mypy has no issue with:
```python
sum("hello")
```
Which results in this error at runtime:
```
TypeError: unsupported operand type(s) for +: 'int' and 'str'
```

context: I was introducing someone to mypy and used the example above as something that mypy should flag, but turns out it wasn't being flagged so logging here.

### PR DESCRIPTION
**Pull request by AlexWaygood**
Closes #7574 (cc. @jpy-git). I think this is just about the best we can do without introducing many false-positive errors.

### CODE REVIEW

File: stdlib/_typeshed/__init__.pyi [hunk 1]
```diff
@@ -69,9 +69,12 @@ SupportsRichComparisonT = TypeVar("SupportsRichComparisonT", bound=SupportsRichC
 
 # Dunder protocols
 
-class SupportsAdd(Protocol):
+class SupportsAddition(Protocol):
```
Comment (hauntsaninja) [inline stdlib/_typeshed/__init__.pyi:72]: (I think I prefer the old name since it mirrors the dunder method name)

Comment (AlexWaygood) [reply stdlib/_typeshed/__init__.pyi:72]: I disliked the idea of `SupportsAddWithInt` ://

But fair enough, I'll revert the naming change :)

**Comment by AlexWaygood** [comment]
Any thoughts from anyone on whether this is a good idea? :)

File: stdlib/builtins.pyi [hunk 2]
```diff
@@ -1561,15 +1562,15 @@ else:
     def sum(__iterable: Iterable[bool], __start: int = ...) -> int: ...  # type: ignore[misc]
 
 @overload
-def sum(__iterable: Iterable[_SumT]) -> _SumT | Literal[0]: ...
+def sum(__iterable: Iterable[_AddableWithIntT]) -> _AddableWithIntT | Literal[0]: ...
```
Comment (srittau) [inline stdlib/builtins.pyi:1565]: This doesn't seem right:

```python
In [2]: class Foo:
   ...:     def __add__(self, other):  # accepts int as "other"
   ...:         return Foo()
   ...: 

In [3]: sum([Foo(), Foo()])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-55dd5d052a18> in <module>
----> 1 sum([Foo(), Foo()])

TypeError: unsupported operand type(s) for +: 'int' and 'Foo'
```

Comment (AlexWaygood) [reply stdlib/builtins.pyi:1565]: That's a weird `TypeError` there though. Good spot. Maybe `__radd__` is involved somehow.

Comment (AlexWaygood) [reply stdlib/builtins.pyi:1565]: Indeed, `__add__` and `__radd__` are both required if no default is specified:

```python
>>> class Foo:
...     def __add__(self, other):
...         return Foo()
...
>>> sum([Foo(), Foo()])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'int' and 'Foo'
>>> class Bar:
...     def __radd__(self, other):
...         return Bar()
...
>>> sum([Bar(), Bar()])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'Bar' and 'Bar'
>>> class Baz:
...     def __add__(self, other):
...         return Baz()
...     def __radd__(self, other):
...         return Baz()
...
>>> sum([Baz(), Baz()])
<__main__.Baz object at 0x000001E71D90BD30>
```

**Review by srittau — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
This just found a bug where we were doing `sum()` on a type that defined `__add__` but not `__radd__`. Thanks!

**Pull request merged**